crates/pathfinder/src/rpc/api.rs

```diff
@@ -1160,6 +1169,48 @@ impl RpcApi {
             contract_address: result.address,
         })
     }
+
+    pub async fn estimate_fee(
+        &self,
+        request: Call,
+        block_hash: BlockHashOrTag,
+    ) -> RpcResult<FeeEstimate> {
+        use crate::cairo::ext_py::GasPriceSource;
+
+        match (
+            self.call_handle.as_ref(),
+            self.shared_gas_price.as_ref(),
+            &block_hash,
+        ) {
+            (Some(h), _, &BlockHashOrTag::Hash(_)) => {
+                // discussed during estimateFee work: when using block_hash use the gasPrice from
+                // the starknet_blocks::gas_price column, otherwise (tags) get the latest eth_gasPrice.
+                h.estimate_fee(request, block_hash, GasPriceSource::PastBlock)
+                    .await
+                    .map_err(Error::from)
+            }
+            (Some(h), Some(g), &BlockHashOrTag::Tag(Tag::Latest)) => {
+                // now we want the gas_price from our hopefully pre-fetched source, it will be the
+                // same when we finally have pending support
+
+                let gas_price = g
+                    .get()
+                    .await
+                    .ok_or_else(|| internal_server_error("eth_gasPrice unavailable"))?;
+
+                h.estimate_fee(request, block_hash, GasPriceSource::Current(gas_price))
+                    .await
+                    .map_err(Error::from)
+            }
+            (Some(_), Some(_), &BlockHashOrTag::Tag(Tag::Pending)) => {
+                Err(internal_server_error("Unimplemented"))
+            }
+            _ => {
```

Comment:
[`call` forwards](https://github.com/eqlabs/pathfinder/blob/24bf184ff55c65c9f3b45177278a3e8de5413bf7/crates/pathfinder/src/rpc/api.rs#L981), while `estimate_fee` does not if the handle is missing.

Reply:
There's a comment for that?

Reply:
This came to my mind when I started thinking about the tests that we have in [the previous comment](https://github.com/eqlabs/pathfinder/pull/388#discussion_r901834289).

I guess both `call` and `estimate_fee` should be consistent with what to do when the handle is `None`.

Reply:
Do you mean it should error out in the call as well?